### PR TITLE
feat(v2): introduce migration_mode, default to v1

### PR DESCRIPTION
**Pull request by noahdietz**
Adds a new flag/attribute `migration_mode` with values `MIGRATION_MODE_UNSPECIFIED` (both client surfaces, v1 surface in samples), `MIGRATING` (both client surfaces, v2 surface in samples), and `NEW_SURFACE_ONLY` (only v2 surface generated/used in samples). All unit and integration tests, except for those explicitly set to a value for testing, will generate with `MIGRATION_MODE_UNSPECIFIED`.

This also reworks `SnippetDetails` into two versions - one for the `V1` surface, and one for the `V2` surface. This backs out a bunch of changes that were made previously to have snippets default to `V2` surface generation pre-`migration_mode`.

So in the unit tests, `Basic` and `RoutingHeaders` are set to `NEW_SURFACE_ONLY` and `MIGRATING` respectively, while in the integration tests `functions` and `securitycenter` are set to the same, respectively.

### CODE REVIEW

**Comment by noahdietz** [comment]
https://github.com/googleapis/gapic-generator-php/pull/606/commits/e8f4d7756fc40bcd537a0b4ca0984fd2a44bb49e has the changes introducing `migration_mode`, while https://github.com/googleapis/gapic-generator-php/pull/606/commits/4b1df8237794288693deff48a51abee07f8ab23e has the `SnippetDetails` split and some diff updates.

File: tests/Unit/ProtoTests/ProtoTest.php [hunk 1]
```diff
@@ -68,7 +71,7 @@ private function runProtoTest(
 
     public function testBasic0(): void
     {
-        $this->runProtoTest('Basic/basic.proto');
+        $this->runProtoTest('Basic/basic.proto', null, null, true, MigrationMode::NEW_SURFACE_ONLY);
```
Comment (bshaffer) [inline tests/Unit/ProtoTests/ProtoTest.php:74]: I see in `Basic/out` that the `BasicClient` and `BasicGapicClient` were removed!

Super-duper nit to add a comment, and a suggestion to use named parameters for readability!
```suggestion
        // test generating the client with only the new surface (no v1 client, v2 samples)
        $this->runProtoTest('Basic/basic.proto', migrationMode: MigrationMode::NEW_SURFACE_ONLY);
```

File: tests/Unit/ProtoTests/ProtoTest.php [hunk 2]
```diff
@@ -107,7 +110,7 @@ public function testGrpcServiceConfig(): void
 
     public function testRoutingHeaders(): void
     {
-        $this->runProtoTest('RoutingHeaders/routing-headers.proto');
+        $this->runProtoTest('RoutingHeaders/routing-headers.proto', null, null, true, MigrationMode::MIGRATING);
```
Comment (bshaffer) [inline tests/Unit/ProtoTests/ProtoTest.php:113]:
```suggestion
        // test generating the client in migration mode (both v1 and v2 clients, but with v2 samples)
        $this->runProtoTest('RoutingHeaders/routing-headers.proto', migrationMode: MigrationMode::MIGRATING);
```

File: tests/Unit/ProtoTests/BasicDiregapic/out/samples/LibraryClient/add_comments.php [hunk 3]
```diff
@@ -40,13 +39,10 @@ function add_comments_sample(string $formattedName): void
 
     // Prepare the request message.
     $comments = [new Comment()];
-    $request = (new AddCommentsRequest())
-        ->setName($formattedName)
-        ->setComments($comments);
```
Comment (bshaffer) [inline tests/Unit/ProtoTests/BasicDiregapic/out/samples/LibraryClient/add_comments.php:45]: This is awesome!!  This doesn't really belong in this PR, but we should consider adding the resource name helpers to the samples. E.g, this would be
```php
$request = AddCommentsRequest::build($formattedName, $comments);
```

**Review by bshaffer — APPROVED** [review]

File: tests/Unit/ProtoTests/ProtoTest.php (hunk 1 above)
Comment (noahdietz) [reply tests/Unit/ProtoTests/ProtoTest.php:74]: Oh my gosh named params bless upppppp

File: tests/Unit/ProtoTests/BasicDiregapic/out/samples/LibraryClient/add_comments.php (hunk 3 above)
Comment (noahdietz) [reply tests/Unit/ProtoTests/BasicDiregapic/out/samples/LibraryClient/add_comments.php:45]: > but we should consider adding the resource name helpers

You mean the request builder methods? Yeah you are right. Separate PR tho fosho, and obviously those are in use when the `V2` client is used for sample gen.
